src/document.rs

```diff
@@ -1,15 +1,71 @@
 //! Abstractions for CMark Documents and Howser Templates.
 
+extern crate env_logger;
 extern crate regex;
 extern crate unicode_segmentation;
 
 use self::regex::Regex;
-use std::collections::HashMap;
-use std::cell::RefCell;
-use errors::{HowserError, HowserResult};
+use self::unicode_segmentation::UnicodeSegmentation;
+use constants::{DITTO_TOKEN, MANDATORY_PROMPT, OPTIONAL_PROMPT, PROMPT_PATTERN};
+use data::ElementType;
 use data::{MatchType, NodeData};
-use doogie::Node;
 use doogie::constants::*;
+use doogie::Node;
+use errors::{HowserError, HowserResult, SpecWarning};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use validator::types_match;
+
+#[derive(Debug)]
+enum LookaheadType {
+    IntegratedLiteral(Node),
+    IntegratedOccupied((Node, MatchType)),
+    IntegratedVacant((Node, MatchType)),
+    DiscreteAnnotated((Node, MatchType)),
+    DiscreteLiteral(Node),
+    Ditto(Node),
+    List(Node),
+    Other(Option<Node>),
+}
+
+impl LookaheadType {
+    fn new(node: Option<Node>) -> HowserResult<Self> {
+        if let Some(node) = node {
+            let match_type = extract_annotation(&node)?;
+            match node.capabilities
+                .get
+                .as_ref()
+                .ok_or(HowserError::CapabilityError)?
+                .get_type()?
+            {
+                NodeType::CMarkNodeList => Ok(LookaheadType::List(node)),
+                NodeType::CMarkNodeParagraph
+                | NodeType::CMarkNodeBlockQuote
+                | NodeType::CMarkNodeCodeBlock => match match_type {
+                    MatchType::Repeatable => Ok(LookaheadType::Ditto(node)),
+                    MatchType::None => Ok(LookaheadType::IntegratedLiteral(node)),
+                    _ => match node.capabilities
+                        .traverse
+                        .as_ref()
+                        .ok_or(HowserError::CapabilityError)?
+                        .first_child()?
+                    {
+                        Some(_) => Ok(LookaheadType::IntegratedOccupied((node, match_type))),
+                        None => Ok(LookaheadType::IntegratedVacant((node, match_type))),
+                    },
+                },
+                NodeType::CMarkNodeHeading | NodeType::CMarkNodeItem => match match_type {
+                    MatchType::Repeatable => Ok(LookaheadType::Ditto(node)),
+                    MatchType::None => Ok(LookaheadType::DiscreteLiteral(node)),
+                    _ => Ok(LookaheadType::DiscreteAnnotated((node, match_type))),
+                },
+                _ => Ok(LookaheadType::Other(Some(node))),
+            }
+        } else {
+            Ok(LookaheadType::Other(None))
```

Comment:
Is there any way of inverting this into an early return? Since this is returning an `Ok`, the `?` operator probably won't work, but maybe a `match` up front with a `return` in the `None` branch?

Reply:
Then I would still have to unwrap the Option though to get to the Node which would still leave me with a dangling logical branch